src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,190 @@
+                {
+                    if (!string.IsNullOrEmpty(stderr.Data))
+                    {
+                        executionContext.Error(stderr.Data);
+                    }
+                };
+
+                var trimChars = new[] { '\\', '/' };
+
+                dropLocation = Path.Combine(dropLocation.TrimEnd(trimChars));
+                downloadFolderPath = downloadFolderPath.TrimEnd(trimChars);
+
+                string robocopyArguments = "\"" + dropLocation + "\" \"" + downloadFolderPath + "\" * /E /COPY:DA /NP /R:3";
+
+                int exitCode = await processInvoker.ExecuteAsync(
+                        workingDirectory: "",
+                        fileName: "robocopy",
+                        arguments: robocopyArguments,
+                        environment: null,
+                        requireExitCodeZero: false,
+                        outputEncoding: null,
+                        killProcessOnCancel: true,
+                        cancellationToken: cancellationToken);
+
+                executionContext.Output(StringUtil.Loc("RobocopyBasedPublishArtifactTaskExitCode", exitCode));
+
+                if (exitCode >= 8)
+                {
+                    throw new Exception(StringUtil.Loc("RobocopyBasedPublishArtifactTaskFailed", exitCode));
+                }
+            }
+        }
+
+        private async Task DownloadFileShareArtifactAsync(string sourcePath, string destPath, int parallelCount, CancellationToken cancellationToken, IEnumerable<Func<string, bool>> minimatchFuncs = null)
+        {
+            var trimChars = new[] { '\\', '/' };
+
+            // If user has specified a relative folder in the drop, change the drop location itself. 
+            sourcePath = Path.Combine(sourcePath.TrimEnd(trimChars));
+
+            IEnumerable<string> files =
+                new DirectoryInfo(sourcePath).EnumerateFiles("*", SearchOption.AllDirectories)
+                    .Select(path => path.FullName);
+
+            var parallelism = new ExecutionDataflowBlockOptions()
+            {
+                MaxDegreeOfParallelism = parallelCount,
+                BoundedCapacity = 2 * parallelCount,
+                CancellationToken = cancellationToken
+            };
+
+            var actionBlock = NonSwallowingActionBlock.Create<string>(
+                action: async file =>
+                {
+                    if (minimatchFuncs == null || minimatchFuncs.Any(match => match(file))) 
+                    {
+                        string tempPath = Path.Combine(destPath, Path.GetRelativePath(sourcePath, file));
+                        this.context.Output(StringUtil.Loc("CopyFileToDestination", file, tempPath));
+                        FileInfo tempFile = new System.IO.FileInfo(tempPath);
+                        using (StreamReader fileReader = GetFileReader(file))
+                        {
+                            await WriteStreamToFile(
```

Comment:
>WriteStreamToFile [](start = 34, length = 17)

I thought for download we will use file copy and not stream?

Reply:
The code you're viewing is outdated I think.